include/pybind11/eigen.h

```diff
@@ -519,9 +554,14 @@ struct type_caster<
         bool need_copy = !isinstance<Array>(src);
 
         EigenConformable<props::row_major> fits;
-        bool is_pyobject = false;
-        if (is_pyobject_<Scalar>()) {
-            is_pyobject = true;
+        constexpr bool is_pyobject = is_pyobject_dtype<Scalar>::value;
+        // TODO(eric.cousineau): Re-enable this once this does not aversely
```

Comment:
I don't understand when we will be enabling the assert?

Reply:
It would be re-enabled once https://github.com/RobotLocomotion/drake/pull/8452 either merges (in which case no code should run across this) or fails (in which case we want to quickly squash these errors).

As of https://github.com/RobotLocomotion/drake/pull/8665, I've only taken basic steps to support multiple scalar types, some of which compiles this code. I want a run-time error now, so that if anybody uses them right now, they get a quick error, but everything still compiles for the high-level API checks on CI.

Once one of the above happens, then it make sense to elevate to compile-time to more quickly squash its usage.

Clarified the comment too.